### PR TITLE
Fix for AST models

### PR DESCRIPTION
**Pull request by vidyasiv**
# What does this PR do?

Fixes # (issue)
- Fix for `MIT/ast-finetuned-speech-commands-v2` with run_audio_classification.py which fails with `AttributeError: 'ASTForAudioClassification' object has no attribute 'freeze_feature_encoder'`
- Typo fix in README
- Added test `test_run_audio_classification_ast-finetuned-speech-commands-v2_multi_card`: 3.50 min to run on Gaudi2

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by ssarkar2** [comment]
@vidyasiv Can you please mention the test here, so we can take a look

**Comment by vidyasiv** [comment]
> @vidyasiv Can you please mention the test here, so we can take a look

- `FAILED tests/test_examples.py::MultiCardAudioClassificationExampleTester::test_run_audio_classification_wav2vec2-base_multi_card - AssertionError: 134 != 0`
- `FAILED tests/test_examples.py::MultiCardSpeechRecognitionExampleTester::test_run_speech_recognition_ctc_wav2vec2-large-lv60_multi_card - AssertionError: 134 != 0`

#### Setup
- Environment: v1.15.0, build 489 
- Commands: 
    - `export GAUDI2_CI=1`
    - `export RUN_SLOW=1`
    - `python -m pip install .[tests]`
    - `python -m pytest tests/test_examples.py -v -s -k "wav2vec2"`

As for the test I am trying to add, I was trying to add it as a wav2vec2 model which I now realize `MIT/ast-finetuned-speech-commands-v2` is not. I will work to add it as audio-spectogram-transformer

File: tests/baselines/ast_finetuned_speech_commands_v2.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
```
Comment (vidyasiv) [inline tests/baselines/ast_finetuned_speech_commands_v2.json:1]: TODO: Need to update values when I can run test, currently segfaults as well

File: tests/test_examples.py [hunk 2]
```diff
@@ -220,6 +220,8 @@ def to_test(model_name: str, multi_card: bool, deepspeed: bool, example_name: st
             return True
         elif "bloom" in model_name and deepspeed and os.environ.get("GAUDI2_CI", "0") == "0":
             return True
+        elif "ast-finetuned-speech-commands-v2" in model_name and os.environ.get("GAUDI2_CI", "0") == "1":
+            return True
```
Comment (regisss) [inline tests/test_examples.py:224]: You need to move this right before line 195, otherwise this test will also be executed on Gaudi1

**Comment by vidyasiv** [comment]
> Apologies, what I told you to disable the test on Gaudi1 was not right! I just pushed a new commit to fix it.
> 
> Besides, when I run the test on Gaudi2 IDC instance, it fails with the following metrics:
> 
> ```
> E   AssertionError: 135.6979 not less than or equal to 121.95645 : for metric train_runtime.                                                 
> E   ===== Assessed metrics (measured vs thresholded baseline) =====
> E   eval_accuracy: 0.1871603260869565 vs 0.18522899999999998
> E   train_runtime: 135.6979 vs 121.95645
> E   train_samples_per_second: 1946.594 vs 2201.5176500000002
> E   eval_samples_per_second: 2848.259 vs 3572.95855
> ```
> 
> Any idea why?

Thanks, perhaps recent commits modified the runtime, will rerun and update. If numbers seem similar we might have to compare environment.

**Comment by vidyasiv** [comment]
> > Apologies, what I told you to disable the test on Gaudi1 was not right! I just pushed a new commit to fix it.
> > Besides, when I run the test on Gaudi2 IDC instance, it fails with the following metrics:
> > ```
> > E   AssertionError: 135.6979 not less than or equal to 121.95645 : for metric train_runtime.                                                 
> > E   ===== Assessed metrics (measured vs thresholded baseline) =====
> > E   eval_accuracy: 0.1871603260869565 vs 0.18522899999999998
> > E   train_runtime: 135.6979 vs 121.95645
> > E   train_samples_per_second: 1946.594 vs 2201.5176500000002
> > E   eval_samples_per_second: 2848.259 vs 3572.95855
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Any idea why?
> 
> Thanks, perhaps recent commits modified the runtime, will rerun and update. If numbers seem similar we might have to compare environment.

@regisss , these are the tested metrics I see (test passes for me: `python -m pytest tests/test_examples.py -v -s -k "test_run_audio_classification_ast-finetuned-speech-commands-v2_multi_card"`)
```bash
  eval_accuracy               =     0.1872
  train_runtime               = 0:02:00.60
  train_samples_per_second    =   2311.196
  eval_samples_per_second     =   3667.233
```

#### Full metrics for test
```
***** train metrics *****
  epoch                       =       10.0
  max_memory_allocated (GB)   =       6.62
  memory_allocated (GB)       =       3.53
  total_memory_available (GB) =      94.62
  train_loss                  =     1.0186
  train_runtime               = 0:02:00.60
  train_samples_per_second    =   2311.196
  train_steps_per_second      =       9.06
[INFO|trainer.py:1753] 2024-04-26 20:41:32,441 >> Using HPU graphs for inference.
[INFO|trainer.py:1773] 2024-04-26 20:41:32,441 >> ***** Running Evaluation *****
[INFO|trainer.py:1775] 2024-04-26 20:41:32,441 >>   Num examples = 5888
[INFO|trainer.py:1778] 2024-04-26 20:41:32,441 >>   Batch size = 64
100%|██████████| 12/12 [00:01<00:00,  8.53it/s]
***** eval metrics *****
  epoch                       =       10.0
  eval_accuracy               =     0.1872
  eval_loss                   =     4.8405
  eval_runtime                = 0:00:03.98
  eval_samples_per_second     =   3667.233
  eval_steps_per_second       =      7.584
  max_memory_allocated (GB)   =       6.62
  memory_allocated (GB)       =       3.73
  total_memory_available (GB) =      94.62
```
#### Host Environment details:
```
HL-SMI Version:                              hl-1.15.0-fw-48.2.1.1  
Driver Version:                                     1.15.0-a596ef0     
```

#### Internal docker image:
```
docker_version: Docker version : 1.15.0
pt_version: PyTorch version : 2.2.0
build_version: Build version :489
HL-SMI Version:                           hl-1.15.0-rc-fw-49.0.0.0
Driver Version:                                     1.15.0-a596ef0  
```

**Comment by regisss** [comment]
Inside the Docker container, I have:
```
HL-SMI Version:                              hl-1.15.0-fw-48.2.1.1                                            
Driver Version:                              1.15.0-a596ef0
Docker image                                 vault.habana.ai/ui/native/gaudi-docker/1.15.0/ubuntu22.04/habanalabs/pytorch-installer-2.2.0/1.15.0-479/
```

And my CPU:
```
Architecture:            x86_64                                                                                                              
  CPU op-mode(s):        32-bit, 64-bit                                                                                                      
  Address sizes:         46 bits physical, 57 bits virtual                                                                                   
  Byte Order:            Little Endian                                                                                                       
CPU(s):                  160                                                                                                                 
  On-line CPU(s) list:   0-159                                                                                                               
Vendor ID:               GenuineIntel                                                                                                        
  Model name:            Intel(R) Xeon(R) Platinum 8380 CPU @ 2.30GHz                                                                        
    CPU family:          6                                                                                                                   
    Model:               106                                                                                                                 
    Thread(s) per core:  2                                                                                                                   
    Core(s) per socket:  40                                                                                                                  
    Socket(s):           2                                                                                                                   
    Stepping:            6                                                                                                                   
    CPU max MHz:         3400.0000                                                                                                           
    CPU min MHz:         800.0000
```
Maybe you have a beefier CPU and that makes data loading faster. What's the output of `lspcu` you get?

**Comment by vidyasiv** [comment]
Rebased branch onto latest main, tried with both docker builds 479 and 489- both pass.
Landed on a different host this time, test still passes:
```
***** train metrics *****
  epoch                       =       10.0
  max_memory_allocated (GB)   =      45.91
  memory_allocated (GB)       =       3.52
  total_memory_available (GB) =      94.62
  train_loss                  =     1.0186
  train_runtime               = 0:01:59.47
  train_samples_per_second    =   2334.724
  train_steps_per_second      =      9.152
***** eval metrics *****
  epoch                       =       10.0
  eval_accuracy               =     0.1872
  eval_loss                   =     4.8405
  eval_runtime                = 0:00:04.00
  eval_samples_per_second     =   3773.397
  eval_steps_per_second       =      7.803
  max_memory_allocated (GB)   =      45.91
  memory_allocated (GB)       =       3.73
  total_memory_available (GB) =      94.62
```

HL-SMI
```
HL-SMI Version:                              hl-1.15.0-fw-48.2.1.1
Driver Version:                                     1.15.0-a596ef0  
```

For lscpu:
```
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         46 bits physical, 57 bits virtual
  Byte Order:            Little Endian
CPU(s):                  160
  On-line CPU(s) list:   0-159
Vendor ID:               GenuineIntel
  Model name:            Intel(R) Xeon(R) Platinum 8380 CPU @ 2.30GHz
    CPU family:          6
    Model:               106
    Thread(s) per core:  2
    Core(s) per socket:  40
    Socket(s):           2
    Stepping:            6
    CPU max MHz:         3400.0000
    CPU min MHz:         800.0000
    BogoMIPS:            4600.00
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good
                          nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsa
                         ve avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjus
                         t bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup
                         _llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni a
                         vx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization features: 
  Virtualization:        VT-x
Caches (sum of all):     
  L1d:                   3.8 MiB (80 instances)
  L1i:                   2.5 MiB (80 instances)
  L2:                    100 MiB (80 instances)
  L3:                    120 MiB (2 instances)
NUMA:                    
  NUMA node(s):          2
  NUMA node0 CPU(s):     0-39,80-119
  NUMA node1 CPU(s):     40-79,120-159
Vulnerabilities:         
  Gather data sampling:  Mitigation; Microcode
  Itlb multihit:         Not affected
  L1tf:                  Not affected
  Mds:                   Not affected
  Meltdown:              Not affected
  Mmio stale data:       Mitigation; Clear CPU buffers; SMT vulnerable
  Retbleed:              Not affected
  Spec rstack overflow:  Not affected
  Spec store bypass:     Mitigation; Speculative Store Bypass disabled via prctl and seccomp
  Spectre v1:            Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:            Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
  Srbds:                 Not affected
  Tsx async abort:       Not affected
```
Feel free to modify values per machine config that will be used for regressions.

**Comment by regisss** [comment]
@vidyasiv I cannot push to your fork, can you update the following baselines please?
```
"train_runtime": 139.9477,
"train_samples_per_second": 1955.74,
"eval_samples_per_second": 2301.088,
```
`eval_samples_per_second` is quite lower compared to yours because I get inconsistent results, sometimes larger than 3000 and sometimes as low as 2300. Not sure why this happens.

**Comment by vidyasiv** [comment]
> @vidyasiv I cannot push to your fork, can you update the following baselines please?
> 
> ```
> "train_runtime": 139.9477,
> "train_samples_per_second": 1955.74,
> "eval_samples_per_second": 2301.088,
> ```
> 
> `eval_samples_per_second` is quite lower compared to yours because I get inconsistent results, sometimes larger than 3000 and sometimes as low as 2300. Not sure why this happens.

Updated. Do you see the inconsistency on the same host?

**Review by regisss — APPROVED** [review]

**Comment by regisss** [comment]
> > @vidyasiv I cannot push to your fork, can you update the following baselines please?
> > ```
> > "train_runtime": 139.9477,
> > "train_samples_per_second": 1955.74,
> > "eval_samples_per_second": 2301.088,
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `eval_samples_per_second` is quite lower compared to yours because I get inconsistent results, sometimes larger than 3000 and sometimes as low as 2300. Not sure why this happens.
> 
> Updated. Do you see the inconsistency on the same host?

Yes, on the same host. Maybe there are re-compilations during evaluation? Even a 1 second difference in a compilation could impact the final throughput quite a lot. The initial compilation that takes place during the first couple of iterations is discarded for computing the throughput though.
I'm going to merge it that way, let's see if I can get a better understanding of what is going on here.

**Pull request merged**